File: year2/systems-design-implementation/Lab4/Lab4Sharp/Repo/Database/ParticipantRepositoryDB.cs
using Lab2Sharp.Domain;
using Lab2Sharp.Repo.Interfaces;
using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;
using System.Threading.Tasks;
using System.Data.SQLite;
using log4net;
using Lab2Sharp.Domain.Validator;

namespace Lab2Sharp.Repo.Database
{
    public class ParticipantRepositoryDB: ParticipantRepository
    {
        private string url;
        private static readonly ILog log = LogManager.GetLogger("ParticipantRepository");
        private ParticipantValidator validator;

        public ParticipantRepositoryDB(string url, ParticipantValidator validator)
        {
            log.Info("Creating ParticipantRepository");
            this.url = url;
            this.validator = validator;
        }

        public Participant FindOne(int id)
        {
            log.InfoFormat("Entering findOne with value {0}", id);
            string sqlCom = "select * from Participanti where id=@id";
            using (var conn = new SQLiteConnection(url))
            {
                conn.Open();
                using (var cmd = new SQLiteCommand(sqlCom, conn))
                {
                    var idd = cmd.CreateParameter();
                    idd.ParameterName = "@id";
                    idd.Value = id;
                    cmd.Parameters.Add(idd);
                    using (var data = cmd.ExecuteReader())
                    {
                        if (data.Read())
                        {
                            String nume = data.GetString(1);
                            int varsta = data.GetInt32(2);
                            Participant p = new Participant(id, nume, varsta);
                            log.InfoFormat("Exiting findOne with value {0}", p);
                            return p;
                        }
                    }
                }
            }
            log.InfoFormat("Exiting findOne with value {0}", null);
            return null;
        }
        public IEnumerable<Participant> FindAll()
        {
            log.InfoFormat("Entering findAll");
            IList<Participant> list = new List<Participant>();
            string sqlCom = "select * from Participanti";
            using (var conn = new SQLiteConnection(url))
            {
                conn.Open();
                var cmd = new SQLiteCommand(sqlCom, conn);
                using (var data = cmd.ExecuteReader())
                {
                    while (data.Read())
                    {
                        int id = data.GetInt32(0);
                        String nume = data.GetString(1);
                        int varsta = data.GetInt32(2);
                        Participant p = new Participant(id, nume, varsta);
                        log.InfoFormat("Found {0}", p);
                        list.Add(p);
                    }
                }
            }
            log.InfoFormat("Exiting findAll");
            return list;
        }
        public Participant Add(Participant entity)
        {
            validator.validate(entity);
            Participant p = null;
            string sqlCom = "insert into Participanti (nume, varsta) values (@nume,@varsta)"; 
            using (var conn = new SQLiteConnection(url)) { 
                conn.Open();
                using (var cmd = new SQLiteCommand(sqlCom, conn))
                {
                    var nume = cmd.CreateParameter();
                    nume.ParameterName = "@nume";
                    nume.Value = entity.nume;
                    cmd.Parameters.Add(nume);
                    var varsta = cmd.CreateParameter();
                    varsta.ParameterName = "@varsta";
                    varsta.Value = entity.varsta;
                    cmd.Parameters.Add(varsta);
                    if (cmd.ExecuteNonQuery() > 0)
                        p = entity;
                }
                sqlCom = "select max(id) as id from Participanti";
                using (var cmd = new SQLiteCommand(sqlCom, conn))
                {
                    using (var data = cmd.ExecuteReader())
                    {
                        if (data.Read())
                        {
                            p.id = data.GetInt32(0);
                        }
                    }
                }
            }
            return p;
        }
        public Participant Delete(int id)
        {
            Participant p = FindOne(id);
            string sqlCom = "delete from Participanti where id=@id";
            using (var conn = new SQLiteConnection(url))
            {
                conn.Open();
                using (var cmd = new SQLiteCommand(sqlCom, conn))
                {
                    var idd = cmd.CreateParameter();
                    idd.ParameterName = "@id";
                    idd.Value = id;
                    cmd.Parameters.Add(idd);
                    cmd.ExecuteNonQuery();
                }
            }
            return p;
        }
        public Participant Update(Participant entity)
        {
            validator.validate(entity);
            Participant p = FindOne(entity.id);
            string sqlCom = "update Participanti set nume=@nume,varsta=@varsta where id=@id";
            using (var conn = new SQLiteConnection(url))
            {
                conn.Open();
                using (var cmd = new SQLiteCommand(sqlCom, conn))
                {
                    var id = cmd.CreateParameter();
                    id.ParameterName = "@id";
                    id.Value = entity.id;
                    cmd.Parameters.Add(id);
                    var nume = cmd.CreateParameter();
                    nume.ParameterName = "@nume";
                    nume.Value = entity.nume;
                    cmd.Parameters.Add(nume);
                    var varsta = cmd.CreateParameter();
                    varsta.ParameterName = "@varsta";
                    varsta.Value = entity.varsta;
                    cmd.Parameters.Add(varsta);
                    cmd.ExecuteNonQuery();
                }
            }
            return p;
        }
    }
}

File: year2/systems-design-implementation/Lab4/Lab4/src/main/java/repo/db/ParticipantRepositoryDB.java
package repo.db;

import domain.Participant;
import domain.validator.ParticipantValidator;
import org.apache.logging.log4j.LogManager;
import org.apache.logging.log4j.Logger;
import repo.intf.ParticipantRepository;

import java.sql.*;
import java.util.ArrayList;
import java.util.List;
import java.util.Properties;

public class ParticipantRepositoryDB implements ParticipantRepository {

    private final static Logger logger = LogManager.getLogger(ParticipantRepositoryDB.class);
    private String url;
    private ParticipantValidator validator;

    public ParticipantRepositoryDB(Properties props, ParticipantValidator validator) {
        this.url = props.getProperty("db.url");
        this.validator = validator;
    }

    @Override
    public Participant findOne(Integer id) {
        logger.traceEntry();
        Participant p = null;
        Connection connection = null;
        try {
            connection = DriverManager.getConnection(url);
        } catch (SQLException throwables) {
            throwables.printStackTrace();
        }
        try (PreparedStatement preparedStatement = connection.prepareStatement("select * from Participanti where id=?")) {
            preparedStatement.setInt(1, id);
            ResultSet resultSet = preparedStatement.executeQuery();
            while (resultSet.next()) {
                String nume = resultSet.getString("nume");
                int varsta = resultSet.getInt("varsta");
                p = new Participant(id, nume, varsta);
            }
        } catch (SQLException ex) {
            logger.error(ex);
            System.err.println(ex);
        }
        logger.traceExit(p);
        return p;
    }

    @Override
    public List<Participant> findAll() {
        logger.traceEntry();
        Connection connection = null;
        try {
            connection = DriverManager.getConnection(url);
        } catch (SQLException throwables) {
            throwables.printStackTrace();
        }
        List<Participant> list = new ArrayList<>();
        try (PreparedStatement preparedStatement = connection.prepareStatement("select * from Participanti")) {
            try (ResultSet resultSet = preparedStatement.executeQuery()) {
                while (resultSet.next()) {
                    int id = resultSet.getInt("id");
                    String nume = resultSet.getString("nume");
                    int varsta = resultSet.getInt("varsta");
                    Participant p = new Participant(id, nume, varsta);
                    list.add(p);
                }
            }
        } catch (SQLException ex) {
            logger.error(ex);
            System.err.println(ex);
        }
        logger.traceExit(list);
        return list;
    }

    @Override
    public Participant add(Participant entity) throws Exception {
        validator.validate(entity);
        logger.traceEntry("Adding task {} ", entity);
        Participant p = null;
        Connection connection = null;
        try {
            connection = DriverManager.getConnection(url);
        } catch (SQLException throwables) {
            throwables.printStackTrace();
        }
        try (PreparedStatement preStmt = connection.prepareStatement("insert into Participanti (nume, varsta) values (?,?)")) {
            preStmt.setString(1, entity.getNume());
            preStmt.setInt(2, entity.getVarsta());
            int result = preStmt.executeUpdate();
            logger.trace("Added {} instance", result);
            p = entity;
        } catch (SQLException ex) {
            logger.error(ex);
            System.err.println("Error DB " + ex);
        }
        try (PreparedStatement preparedStatement = connection.prepareStatement("select max(P.id) as id from Participanti P")) {
            try (ResultSet resultSet = preparedStatement.executeQuery()) {
                while (resultSet.next()) {
                    p.setId(resultSet.getInt("id"));
                }
            }
        } catch (SQLException ex) {
            logger.error(ex);
            System.err.println(ex);
        }
        logger.traceExit();
        return p;
    }

    @Override
    public Participant delete(Integer id) {
        logger.traceEntry();
        Participant p = findOne(id);
        Connection connection = null;
        try {
            connection = DriverManager.getConnection(url);
        } catch (SQLException throwables) {
            throwables.printStackTrace();
        }
        try (PreparedStatement preparedStatement = connection.prepareStatement("delete from Participanti where id=?")) {
            preparedStatement.setInt(1, id);
            preparedStatement.executeUpdate();
            logger.trace("Removed {} instance", p);
        } catch (SQLException ex) {
            logger.error(ex);
            System.err.println(ex);
        }
        logger.traceExit();
        return p;
    }

    @Override
    public Participant update(Participant entity) throws Exception {
        validator.validate(entity);
        logger.traceEntry();
        Participant p = findOne(entity.getId());
        Connection connection = null;
        try {
            connection = DriverManager.getConnection(url);
        } catch (SQLException throwables) {
            throwables.printStackTrace();
        }
        try (PreparedStatement preparedStatement = connection.prepareStatement("update Participanti set nume=?,varsta=? where id=?")) {
            preparedStatement.setString(1, entity.getNume());
            preparedStatement.setInt(2, entity.getVarsta());
            preparedStatement.setInt(3, entity.getId());
            preparedStatement.executeUpdate();
            logger.trace("Updated {} instance", p);
        } catch (SQLException ex) {
            logger.error(ex);
            System.err.println(ex);
        }
        logger.traceExit();
        return p;
    }
}
